crypto/evp/evp_lib.c

```diff
@@ -148,6 +155,50 @@ int EVP_CIPHER_set_asn1_iv(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
     }
     return i;
 }
+
+int evp_cipher_get_asn1_aead_params(EVP_CIPHER_CTX *c, ASN1_TYPE *type)
+{
+    int i = 0;
+    long tl;
+    unsigned char iv[EVP_MAX_IV_LENGTH];
+
+    if (type == NULL)
+        return i;
+
+    i = asn1_type_get_octetstring_int(type, &tl, NULL, EVP_MAX_IV_LENGTH);
+    if (i <= 0)
+        return -1;
+    asn1_type_get_octetstring_int(type, &tl, iv, i);
+
+    if (EVP_CIPHER_CTX_ctrl(c, EVP_CTRL_AEAD_SET_IVLEN, i, NULL) <= 0
```

Comment:
This is not the normal flow for setting the iv. This is normally done through a call to init and pass the other params as NULL. If it is possible I would prefer it worked that way. The flow is already complex enough without adding extra controls. Let me know if that works or not.

Reply:
I did it mainly because of the flow in `cms_EncryptedContent_init_bio` which is for decryption like this:

```
EVP_CipherInit_ex(ctx, ciph, NULL, NULL, NULL, enc)
EVP_CIPHER_asn1_to_param(...)
EVP_CipherInit_ex(ctx, NULL, NULL, ec->key, piv, enc) /* piv is NULL in this case */
```

The thing is that iv is retrieved from ASN1 inside `EVP_CIPHER_asn1_to_param` so I would need to remember it somewhere but that seemed a bit less clean than using ctrl. I might need to rethink though. If you have any suggestion what you think would be the cleanest way, happy to look into that!

Reply:
I was suggesting calling EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc) inside EVP_CIPHER_asn1_to_param(...). Or does that not work because you need a key first??

Reply:
Key usually comes from `CMS_EncryptedContentInfo`. Are you suggesting to use
`EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc)` inside `EVP_CIPHER_asn1_to_param` and then call it again in `cms_EncryptedContent_init_bio` with the key? Basically something like

```
EVP_CipherInit_ex(ctx, ciph, NULL, NULL, NULL, enc) // in cms_EncryptedContent_init_bio
EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc) // in EVP_CIPHER_asn1_to_param
EVP_CipherInit_ex(ctx, NULL, NULL, key, NULL, enc) //  in cms_EncryptedContent_init_bio
```

Also EVP_CIPHER_asn1_to_param is a public function so could it change the ctx behavior somehow? I mean if it will be always the case that `EVP_CipherInit_ex(ctx, NULL, NULL, NULL, piv, enc)` will only set `iv` and doesn't do another initialization.

Reply:
Ahh yeah so that was a stupid idea :)
Ok another idea. Because this is new code that will not be in 1_1_1.
The new aes_gcm code will now generate an IV if you dont specify one (it was required for FIPS). So you shouldnt need to generate it yourself.. The only issue then would be if you need to get the iv back out in order to store it somewhere.. Try taking out the set_iv and see what happens.

Reply:
I'm not sure if I follow. :) The described flow and this line of code is used just for decryption. In that case the iv comes from the ASN.1 params where it was stored when eveloped data has been created. How could it work if I take it out?

Reply:
I'm not sure if I follow. :) The described flow and this line of code is used just for decryption. In that case the iv comes from the ASN.1 params where it was stored when eveloped data has been created. How could it work if I take it out?

Reply:
Ok. I am talking about the set_iv ctrl that you added in the encrypt.. (obviously I need to go to sleep soon)

Reply:
The EVP_CTRL_AEAD_SET_IVLEN should be ok as it is an existing control (I am just looking at the 2 new ctrl id's you added in this PR). one is replaced by the PR above (to get the tag length) and the other one that sets the iv should be handled internally in openssl now.

Reply:
Sorry for the confusion.

Reply:
Ah yeah sorry I actually meant the code line below (EVP_CTRL_SET_IV) :). Anyway that's still used for decryption. Basically the decryption flow is `cms_EncryptedContent_init_bio -> if decryption -> EVP_CIPHER_asn1_to_param -> if GCM -> evp_cipher_get_asn1_aead_params`. This function parses the ASN.1, get the IV and set its length (this line) and then iv itself (line below). So basically I still need to set iv somehow :)

The encryption part is in `evp_cipher_set_asn1_aead_params` which sets tag length (EVP_CTRL_AEAD_GET_TAGLEN) and get iv (EVP_CTRL_GET_IV) so it can be stored in ASN.1.

Reply:
I will have to think about this a bit more. The interface is already confusing enough without yet another way of setting the iv into the ctx.

Reply:
Another option would be to implement it directly on the cipher using `set_asn1_parameters` resp. `get_asn1_parameters`. Then modifying of the context could be a bit more specific but could also result in duplicating the same logic when support for other AEAD ciphers is added in the future. I would probably have to think about it a bit more... Do you think that this would be potentially cleaner?

Reply:
I have done a different approach and wrap the functions so no need for ctrls.